flax/struct.py

```diff
@@ -133,3 +135,20 @@ def from_state_dict(x, state):
 
 def field(pytree_node=True, **kwargs):
   return dataclasses.field(metadata={'pytree_node': pytree_node}, **kwargs)
+
+
+TNode = TypeVar('TNode', bound='PyTreeNode')
+
+
+class PyTreeNode():
+
+  def __init_subclass__(cls):
+    dataclass(cls)  # pytype: disable=wrong-arg-types
+
+  def __init__(self, *args, **kwargs):
+    # stub for pytype
+    raise NotImplementedError
+
+  def replace(self: TNode, **overrides) -> TNode:
```

Comment:
Why do you use `TNode` instead of simply `PyTreeNode` here?

Reply:
Because that's not the correct type. The returned value has the same type as self.
So for example optimizer.replace() returns an Optimizer instance not just an arbitrary PyTreeNode.